_posts/2018-09-06-stateless-microservices.md

```diff
@@ -0,0 +1,304 @@
+
+Let's look at some background before moving into the tutorial showing how to put this into practice.
+
+## What's a Function?
+
+Before getting into what a "stateless microservice" is, let's go back to my definition of a "FaaS Function" as defined in my blog post: [Functions as a Service (FaaS) (Jan 2017)](https://blog.alexellis.io/functions-as-a-service/).
+
+Functions tend to involve:
+
+* invoking short-lived functions (Lambda has a default 1-sec timeout)
+* does not publish TCP services - often accesses third-party services or resources
+* tends to be ad-hoc/event-driven such as responding to webhooks
+* should deal gracefully with spikes in traffic
+* despite name, runs on servers backed by a fluid, non-deterministic infrastructure
+* when working well makes the complex appear simple
+* has several related concerns: infrastructure management, batching, access control, definitions, scaling & dependencies
+
+Since the original post I re-wrote my initial observations as a series of properties.
+
+Functions are:
+
+* stateless
+* emphemeral & short-lived
+* auto-scaling
+* single-purposed
+
+Functions are stateless because they do not rely on internal state-machines, stored files or sessions. Each call to the function should result in the same end-result. They don't have to be strictly idempotent, but they are easier to manage when they are.
+
+Functions are emphemeral because at any point they should be able to be replaced by any similar variant. This property means that all functions can be managed in the same way - from their health-check to their lifecycle, logging and monitoring.
+
+Functions auto-scale either from a minimum replica count to a maximum count, or even down to zero and back again to match demand or to conserve resources.
+
+Functions are single purposed, but common sense must prevail. We do not need to create a new FaaS function every time we write `function x() {}` or `def x:` in our code. A single purpose may be "CRUD for an employee record", "format this IBAN" or "identify the contents of this image using a machine learning model".
+
+If you've written at least one serverless function then you'll be aware that the entrypoint is often abstracted away from you. You just write a handler and you're done - any dependencies are then expressed in a common format such as a `package.json` for Node.js or a `requirements.txt` file for Python.
```

Comment:
Yes. The difference between config and code is interesting, I've always thought. Both need maintaining.